### ISSUE TITLE
pLaTeX crashes with a section contains endash

### ISSUE THREAD

**Issue by tk0miya** [sphinx-doc/sphinx#3181]
In Japanese docs, pLaTeX crashes with a section contains emdash
```
EN DASH:  –
-------------
```
```
$ make clean latexpdfja
...
LaTeX Font Warning: Font shape `JT1/mc/bx/it' undefined
(Font)              using `JT1/mc/bx/n' instead on input line 143.


LaTeX Font Warning: Font shape `JY1/mc/bx/it' undefined
(Font)              using `JY1/mc/bx/n' instead on input line 143.

(/usr/local/texlive/2016/texmf-dist/tex/latex/amsfonts/umsa.fd)
(/usr/local/texlive/2016/texmf-dist/tex/latex/amsfonts/umsb.fd) [1] [2]
(./test.toc
! Text line contains an invalid character.
l.1 ...sline {chapter}{\numberline {1}EN DASH: ^^@
                                                  ^^80^^93}{1}{chapter.1}
```

**Comment by tk0miya** [comment]
Fixed at 69820e1.
I replaced the EM DASH character by `\textemdash{}` macro.

**Comment by jfbu** [comment]
I know nothing about Japanese and platex. But I am wondering if some set-up related to platex is missing. Your example is with EN DASH (U+2013). If I insert an EM DASH (U+2014) I notice that platex appears to convert it to HORIZONTAL BAR (U+2015) in aux and toc file (but might be after-effect if platex is not at all unicode aware). I tried with `jreport` class, it then seems ok in pdf output. But there is problem with EN DASH, why is that (I tried ``platex -etex`` but it doesn't change anything). 

I tried``uplatex`` with ``utarticle`` class. The EN DASH and EM DASH end up as is in aux files. The EN DASH seems ok in pdf but the EM DASH ends up rotated it seems with respect to text.

**Comment by tk0miya** [comment]
Thank you for your help :-)

The unicode support of pLaTeX is not complete. pLaTeX uses EUC-JP or Shift-JIS as a internal encoding. On my local, EUC-JP is used to internal encoding. But, unfortunately, EN DASH is not contained in the character set of EUC-JP (a.k.a JIS X 0208; JIS means Japan Industrial Standard).
I guess this is why the EN DASH is broken on aux file.

Note: EM DASH is contained in JIS X 0208. so it is not broken.

>I tried uplatex with utarticle class. The EN DASH and EM DASH end up as is in aux files. The EN DASH seems ok in pdf but the EM DASH ends up rotated it seems with respect to text.

`uplatex` uses utf-8 encoding as a internal encoding. So it does not convert the text internally. So it helps to compile the document successfully.

Of course, It would be nice to use `uplatex` for Japanese LaTeX users by default. But, to do that, we have to work a little to migration. So I would like to replace EN DASH to `\textendash` on 1.5 series. AFAIK It does not harm nobody.
What do you think?

**Comment by jfbu** [comment]
Thanks for very clear explanations. As far as I can tell having Sphinx LaTeX output use `\textendash` for the Unicode EN DASH (U+2013) should not harm Sphinx users: when using `[utf8]{inputenc}`, the EN DASH character becomes active and expands to `\textendash` (as one can see doing `$ grep textendash *` and `$ grep 2013 *` in ``/usr/local/texlive/2016/texmf-dist/tex/latex/base``). One can imagine macros requiring use of explicit EN DASH character but this is a worry more for general LaTeX, not for LaTeX as one target of Sphinx.

For xelatex engine + fontspec, if one does ``$ grep endash *`` in ``/usr/local/texlive/2016/texmf-dist/tex/latex/fontspec`` one gets

    tufontrange-T1.def:\DeclareUnicodeSymbol{\textendash}{"2013}

hence one sees the macro expands back to the suitable Unicode code-point.

LaTeX's inputenc maps U+2012 (the FIGURE DASH ‒) to `\textendash` for T1 and utf8 encodings, thus it seems to not distinguish the two.

I am not sure I understand why you need to replace the U+2013 by `\textendash` to support migration to `uplatex`, as the latter actually supports Unicode ? it llooks to me more like a patch to circumvent incompatibility of `platex` with Unicode ? in that case one can not push the logic too far, because it might be at times useful to leave the Unicode characters as is in TeX file, if only because they take less bytes than macro names.

**Comment by jfbu** [comment]
as an alternative to have Sphinx systematically replace U+2013 by `\textendash` one could add, only for `platex` engine some special code:

```latex
\makeatletter
\edef\SphinxEII{\string^^e2}
\catcode`\^^e2\active
\def^^e2#1#2%
 {\expandafter\SphinxCheck\csname u8:\SphinxEII\string#1\string#2\endcsname}
\def\SphinxCheck #1{\ifx #1\relax
  \typeout{ERROR: unknown character \expandafter\@gobblefour\string#1}%
  \else\expandafter#1\fi}
\expandafter\def\csname u8:\SphinxEII^^80^^93\endcsname{\textendash}
\makeatother
```
This imitates what utf8 does. It assumes of course that document does not load utf8 package.

**Comment by jfbu** [comment]
caveat: no idea if the code of my previous comment (which should end up either in document preamble or in sphinx.sty) causes any harm to Japanese documents... I tried a bit and it seems ok. The Error message does not raise an error actually, but one could use `\PackageWarning`

**Comment by tk0miya** [comment]
>I am not sure I understand why you need to replace the U+2013 by \textendash to support migration to uplatex, as the latter actually supports Unicode ? 

No, the replacement is introduced only for platex. uplatex supports unicode completely, so it is not needed.

**Comment by jfbu** [comment]
I misunderstood. I thought https://github.com/sphinx-doc/sphinx/commit/69820e1203d7b5b6b7285e3d12826d0bbbf4e06a did a replacement applying to all engines.

**Comment by tk0miya** [comment]
Wait, the replacement works to all engines.
I meant: 
* At this moment, Sphinx uses pLaTeX by default for Japanese docs. And it causes errors if document includes EN DASH
* upLaTeX doesn't have such problem. so to use it is truely solution. BUT sphinx does not support it. To use it, we have to improve settings more.
* I think the replacement does not harm sphinx users.
* Based on these, the replacement is better solution for this case, I believe

**Comment by tk0miya** [comment]
in addition, my comment above means
* the replacement is introduce to fix the issue of pLaTeX
* upLaTeX does not have such problem, so this is not needed for the engine if we will switch to it in future

**Comment by jfbu** [comment]
ok. I agree this replacement to fix issue of pLaTeX should cause no harm to Sphinx users in general, for memo I proposed at https://github.com/sphinx-doc/sphinx/issues/3181#issuecomment-264654543 specific patch which could be applied (in improved form regarding error message) only in case of pLaTeX and avoid having Sphinx escape u+2013 to `\textendash` for all engines. However I can't test it extensively, lacking knowledge of Japanese ;-) I propose I make a PR for future reference.

**Comment by tk0miya** [comment]
Okay, I will test it on Japanese docs.
Thank you for reviewing in detail!

**Comment by jfbu** [comment]
alternative PR is #3193 for future perhaps once it is made better (see comments there)

in the meantime, agreed about https://github.com/sphinx-doc/sphinx/commit/7fccbea1ae96e3db3ce1bdaa8268b07a6e5aacf0 having fixed this issue !

**Comment by jfbu** [comment]
It seems U+2013 (EN DASH) `–` works now with platex (TL2019). (I simply inserted it in a Sphinx project and tried with `language = 'ja'`).

More precisely using the U+2013 or the `\textendash{}` even in section titles seem to behave the same. They both make `dvipdfmx` complain this way:
```
dvipdfmx:warning: No character mapping available.
 CMap name: EUC-UCS2
 input str: <85>
```
which I believe is related to the PDF bookmarks. The PDF text itself is ok, and copying paste works and does give back U+2013, the problem is with the PDF bookmark.

Somehow I feel I am duplicating an issue I already raised... indeed! #4187
